src/core/include/units/magnitude.h

```diff
@@ -0,0 +1,360 @@
+
+constexpr bool is_valid_base_power(const BasePower auto &bp) {
+  if (bp.power == 0) { return false; }
+
+  if constexpr (std::is_same_v<decltype(bp.get_base()), int>) { return is_prime(bp.get_base()); }
+  else { return bp.get_base() > 0; }
+}
+
+// A function object to apply a predicate to all consecutive pairs of values in a sequence.
+template<typename Predicate>
+struct pairwise_all {
+  Predicate predicate;
+
+  template<typename... Ts>
+  constexpr bool operator()(Ts&&... ts) const {
+    // Carefully handle different sizes, avoiding unsigned integer underflow.
+    constexpr auto num_comparisons = [](auto num_elements) {
+      return (num_elements > 1) ? (num_elements - 1) : 0;
+    }(sizeof...(Ts));
+
+    // Compare zero or more pairs of neighbours as needed.
+    return [this]<std::size_t... Is>(std::tuple<Ts...> &&t, std::index_sequence<Is...>) {
+      return (predicate(std::get<Is>(t), std::get<Is + 1>(t)) && ...);
+    }(std::make_tuple(std::forward<Ts>(ts)...), std::make_index_sequence<num_comparisons>());
+  }
+};
+
+// Deduction guide: permit constructions such as `pairwise_all{std::less{}}`.
+template<typename T>
+pairwise_all(T) -> pairwise_all<T>;
+
+// Check whether a sequence of (possibly heterogeneously typed) values are strictly increasing.
+template<typename... Ts>
+  requires ((std::is_signed_v<Ts> && ...))
+constexpr bool strictly_increasing(Ts&&... ts) {
+  return pairwise_all{std::less{}}(std::forward<Ts>(ts)...);
+}
+
+} // namespace detail
+
+/**
+ * @brief  A representation for positive real numbers which optimizes taking products and rational powers.
+ *
+ * Magnitudes can be treated as values.  Each type encodes exactly one value.  Users can multiply, divide, raise to
+ * rational powers, and compare for equality.
+ */
+template<BasePower auto... BPs>
+  requires ((detail::is_valid_base_power(BPs) && ... && detail::strictly_increasing(BPs.get_base()...)))
+struct magnitude {};
+
+// Implementation for Magnitude concept (below).
+namespace detail {
+template<typename T>
+static constexpr bool is_magnitude = false;
+template<BasePower auto... BPs>
+static constexpr bool is_magnitude<magnitude<BPs...>> = true;
+} // namespace detail
+
+/**
+ * @brief  Concept to detect whether T is a valid Magnitude.
+ */
+template<typename T>
+concept Magnitude = detail::is_magnitude<T>;
+
+/**
+ * @brief  Convert any positive integer to a Magnitude.
+ *
+ * This will be the main way end users create Magnitudes.  They should rarely (if ever) create a magnitude<...> by
+ * manually adding base powers.
+ */
+template<ratio R>
+  requires (R.num > 0)
+constexpr Magnitude auto as_magnitude();
+
+/**
+ * @brief  A base to represent pi.
+ */
+struct pi_base {
+  static constexpr long double value = std::numbers::pi_v<long double>;
+};
+
+/**
+ * @brief  A simple way to create a Magnitude representing a rational power of pi.
+ */
+template<ratio Power>
+constexpr auto pi_to_the() { return magnitude<base_power<pi_base>{Power}>{}; }
```

Comment:
I am not sure if we need that. This will not be used too often in the library (only for angles). Maybe moved to unit tests file if needed.